network_simulator/flow.py

```diff
@@ -89,5 +92,218 @@ def handle_packet_success(self, packet_id):
+                            self.flow.source_addr, self.flow.dest_addr,
+                            curr_time)
+        return packet
+
+    def run(self, main_event_loop, statistics):
+        """
+        Set up a list of data packets to initiate a flow.
+
+        :param MainEventLoop main_event_loop: main loop to retrieve the
+        simulation time of function call.
+        :param Statistics statistics: the Statistics to update.
+        """
+        for curr_packet in self.packet_ids:
+            if not self.flow.packet_id_exceeds_data(curr_packet):
+                new_data_packet = self.create_data_packet(
+                    curr_packet, main_event_loop.global_clock_sec)
+                self.packets_to_send.append(new_data_packet)
+
+    def schedule_new_events(self, main_event_loop):
+        """
+        Schedule a FlowSendPacketsEvent immediately, with the list of packets
+        to send.
+
+        :param MainEventLoop main_event_loop: event loop where new Events will
+        be scheduled.
+        """
+        flow_send_event = FlowSendPacketsEvent(self.flow,
+                                               self.packets_to_send)
+        main_event_loop.schedule_event_with_delay(flow_send_event, 0.0)
+
+
+
+class FlowSendPacketsEvent(Event):
+    def __init__(self, flow, packets_to_send):
+        """
+        Send data packet from a flow.
+
+        :ivar Flow flow: flow of this event.
+        :ivar list packets_to_send: list of data packets to send.
+        """
+        self.flow = flow
+        self.packets_to_send = packets_to_send
+
+    def run(self, main_event_loop, statistics):
+        """
+        Check that the size of packets combined is smaller than
+        the window size.
+        :param MainEventLoop main_event_loop: main loop.
+        :param Statistics statistics: the Statistics to update.
+        """
+        if len(set(self.flow.packets_in_transit) \
+            + set(self.flow.packets_to_send)) \
+            > self.flow.window_size_packets:
+                # TODO(team): What do we want to throw here?
+                pass
+
+        for curr_packet in self.packets_to_send:
+            if curr_packet.packet_id not in self.flow.packets_in_transit:
+                self.flow.packets_in_transit.add(curr_packet.packet_id)
+
+    def schedule_new_events(self, main_event_loop):
+        """
+        Schedule DeviceToLinkEvent and renew FlowTimeoutPacketEvent for each
+        packet.
+
+        :param MainEventLoop main_event_loop: event loop where new Events will
+        be scheduled.
+        """
+        # TODO(sharon): Need to fix the initialization after Link subclasses
+        # have been implemented.
+        for curr_packet in self.packets_to_send:
+            device_link_event = DeviceToLinkEvent(curr_packet, False)
+            main_event_loop.schedule_event_with_delay(device_link_event, 0.0)
+
+        # Go through the queue and remove any previously scheduled
```

Comment:
Just schedule another timeout event. If we get to it and there's no need to do anything, then we'll just skip over it.